### PR TITLE
Add Knight#valid_moves

### PR DESCRIPTION
**Pull request by BryceCreamCone**
[TR45] https://trello.com/c/iTgpe9A2/45-front-end-implement-move-restriction 
Wanted a user to get an alert when they made an invalid move
The server was not responding initially, it was being handled on the front end but not working correctly.

[TR47] https://trello.com/c/X8DCbH7Y/47-post-review-cleanup
Removed some code duplication by creating an IconLibrary Class to put image onto Piece#image
Fixed board (squares were the opposite colors of what they should have been)
Cleaned up games - show.html.erb to remove unused Javascript

[TR29] https://trello.com/c/qq0hrJT5/29-constrain-knight-to-valid-moves
Implemented Knight#valid_moves

### CODE REVIEW

File: app/models/game.rb [hunk 1]
```diff
@@ -7,6 +7,23 @@ class Game < ApplicationRecord
   has_many :queens
   has_many :pieces
   scope :available, -> { where(black_player_id: nil) }
+  attr_accessor :turn_counter
```
Comment (CCrowRun) [inline app/models/game.rb]: This may be where you looked at server side turn tracking

File: app/models/icon_library.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class IconLibrary
```
Comment (CCrowRun) [inline app/models/icon_library.rb]: Oooh I like this!

File: app/models/rook.rb [hunk 3]
```diff
@@ -1,13 +1,9 @@
...
   belongs_to :game
 
-  def image
-    return self.image = "&#9814" if self.is_white_piece?
-    self.image = "&#9820"
-  end
-
   def valid_moves
     @valid_moves = []
+    #@valid_moves ||= []
```
Comment (CCrowRun) [inline app/models/rook.rb]: Is this line still needed?

File: app/views/games/show.html.erb [hunk 4]
```diff
@@ -2,13 +2,15 @@
...
-  let activePlayer = "white";
-  let inactivePlayer = "black";
+  let activePlayer = ( setActivePlayer = () => {
+    if (localStorage.getItem('ActivePlayer') == null) {
+      localStorage.setItem('ActivePlayer', 'white');
+    }
+    return localStorage.getItem('ActivePlayer');
+  })();
 
-  //Setup Promise to Grab board
```
Comment (CCrowRun) [inline app/views/games/show.html.erb]: I believe these comments helped with readability - is this something we're going to rethink? Or just remove all comments?

File: app/views/games/show.html.erb [hunk 5]
```diff
@@ -100,25 +93,21 @@ document.addEventListener("DOMContentLoaded", function(){
     initialSquare.removeAttribute("data-piece-id");
     initialSquare.removeEventListener("click", updateActivePiece);
 
-    [activePlayer, inactivePlayer] = [inactivePlayer, activePlayer];
+    change_turns();
```
Comment (CCrowRun) [inline app/views/games/show.html.erb]: I like this means of handling it a lot!

File: spec/controllers/games_controller_spec.rb [hunk 6]
```diff
@@ -91,15 +91,5 @@
       expect(game.black_player_id).to eq(user2.id)
       expect(player2.username).to eq(user2.username)
     end
-
-    it "will not allow two players with the same ID" do
```
Comment (CCrowRun) [inline spec/controllers/games_controller_spec.rb]: Are we removing this as a requirement? Or are we testing this elsewhere?

Comment (BryceCreamCone) [reply spec/controllers/games_controller_spec.rb]: Yup! Before we didn't have a great way to handle it, but I tweaked the controller a little. We should let users access their own games as long as they are active. They won't be added as the other player at all. They'll just load right back into their game.

File: app/views/games/show.html.erb (hunk 4 above)
Comment (BryceCreamCone) [reply app/views/games/show.html.erb]: If you preferred them we can add them back in. It just felt like most of our names had enough intent behind them when I was going through, especially now that it is cleaned up.

File: app/models/rook.rb (hunk 3 above)
Comment (BryceCreamCone) [reply app/models/rook.rb]: Oops. Nope, it is not

File: app/models/icon_library.rb (hunk 2 above)
Comment (BryceCreamCone) [reply app/models/icon_library.rb]: Thanks! It worked even better than I thought too! I was able to move the image selecting into a method in Piece.rb itself, and it should be super easy to change the images now if we'd like!

File: app/models/game.rb (hunk 1 above)
Comment (BryceCreamCone) [reply app/models/game.rb]: Yes, remembering it now I think the server does do it. My real issue was trying to get JavaScript to read it as well. So I misremembered that this morning :) Sorry if that caused some confusion!

Comment (BryceCreamCone) [reply app/models/game.rb]: Yeah, I made all the methods work with Game, but I was having a really hard time having JavaScript read it. There's likely a solution without JavaScript, but the pure JavaScript solution came to me faster. We can either look into implementing these methods or just delete them from Game.rb

File: app/views/games/show.html.erb (hunk 4 above)
Comment (CCrowRun) [reply app/views/games/show.html.erb]: That works for me! I'll mark this as resolved

File: spec/controllers/games_controller_spec.rb (hunk 6 above)
Comment (CCrowRun) [reply spec/controllers/games_controller_spec.rb]: Awesome! Sounds good!

File: app/models/game.rb (hunk 1 above)
Comment (CCrowRun) [reply app/models/game.rb]: Since we have a functioning solution on the JavaScript side, let's remove these from the controller!

**Pull request merged**